lib/Containers/MerkleTree.cpp

```diff
@@ -635,40 +836,46 @@ void MerkleTree<Hasher, BranchingBits, LockStripes>::serialize(velocypack::Build
   velocypack::ArrayBuilder nodeArrayGuard(&output, StaticStrings::RevisionTreeNodes);
   for (std::uint64_t index = 0; index < nodeCountUpToDepth(depth); ++index) {
     velocypack::ObjectBuilder nodeGuard(&output);
-    std::unique_lock<std::mutex> guard(this->lock(index));
-    Node& node = this->node(index);
+    Node const& node = this->node(index);
     output.add(StaticStrings::RevisionTreeHash,
                basics::HybridLogicalClock::encodeTimeStampToValuePair(node.hash, ridBuffer));
     output.add(StaticStrings::RevisionTreeCount, velocypack::Value(node.count));
   }
 }
 
-template <typename Hasher, std::uint64_t const BranchingBits, std::uint64_t const LockStripes>
-void MerkleTree<Hasher, BranchingBits, LockStripes>::serializeBinary(std::string& output,
-                                                                     bool compress) const {
-  std::unique_lock<std::shared_mutex> guard(_bufferLock);
+template <typename Hasher, std::uint64_t const BranchingBits>
+void MerkleTree<Hasher, BranchingBits>::serializeBinary(std::string& output,
+                                                        bool compress) const {
+  std::shared_lock<std::shared_mutex> guard(_bufferLock);
+
   TRI_ASSERT(output.empty());
+    
   if (compress) {
-    snappy::Compress(reinterpret_cast<char*>(_buffer.get()),
-                     allocationSize(meta().maxDepth), &output);
-    output.push_back(::CompressedCurrent);
+    if (this->node(0).count <= 15'000) {
+      // 15'000 is an arbitrary cutoff value for when to move from
+      // bottom-most level compression to full tree compression with Snappy
```

Comment:
15k seems pretty low to me. With depth 6 the bottom-most level has ~262k nodes. Even if every document is assigned to a different node, only ~6% of the nodes have any data.

Reply:
Yes. But at 15k nodes to store, we would need 15kb x 20 bytes, i.e. 300kb already.
Snappy will compress this to about the same size, and is potentially faster than our own implementation. But definitely the cut-off value is very arbitrary.